### PR TITLE
Subquery pushdown - Unions Phase - 2

### PR DESCRIPTION
**Pull request by onderkalaci**
(The PR currently is based on a branch which is not rebased to its parent branch. Thus, we see some older commits. As we progress to review this patch, I'll rebase all PRs accordingly):

    Add top level union support
    
    This commit adds support for UNION/UNION ALL subqueries that are
    in the following form:
    
         .... (Q1 UNION Q2 UNION ...) as union_query ....
    
    In other words, Citus supports allow top level
    unions being wrapped into aggregations queries
    and/or simple projection queries that only selects
    some fields from the lower level queries.

### CODE REVIEW

File: src/backend/distributed/planner/multi_planner.c [hunk 1]
```diff
@@ -117,6 +117,8 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 
 		if (needsDistributedPlanning)
 		{
+			relationRestrictionContext->parseTree = copyObject(originalQuery);
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_planner.c]: I think we should move this line to `CreateAndPushPlannerContexts ()` to make it more readable

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 2]
```diff
@@ -2832,9 +2832,9 @@ ErrorIfContainsUnsupportedSubquery(MultiNode *logicalPlanNode,
 	/* currently in the planner we only allow one subquery in from-clause*/
 	Assert(list_length(subqueryMultiTableList) == 1);
 
-	restrictionEquivalenceForPartitionKeys =
-		RestrictionEquivalenceForPartitionKeys(plannerRestrictionContext);
-	if (!restrictionEquivalenceForPartitionKeys)
+	safeToPushdownSubquery =
+		SafeToPushDownSubquery(plannerRestrictionContext);
+	if (!safeToPushdownSubquery)
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: I don't like this generic error message. We should differentiate unions and joins, at least,

File: src/backend/distributed/planner/relation_restriction_equivalence.c [hunk 3]
```diff
@@ -1052,3 +1259,93 @@ AttributeEquivalancesAreEqual(AttributeEquivalenceClass *firstAttributeEquivalan
...
+static bool
+TopLevelUnionQuery(Query *queryTree)
+{
+	List *rangeTableList = queryTree->rtable;
+	Node *setOperations = queryTree->setOperations;
+	RangeTblEntry *rangeTableEntry = NULL;
+	Query *subqueryTree = NULL;
+
+	/* don't allow joines on top of unions */
+	if (list_length(rangeTableList) > 1)
```
Comment (mtuncer) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: I am not following this. 
`query->rtable `has nothing to do with joins.  joined range table entries can be found from iterating through `query->tree`

Comment (mtuncer) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: I wonder if we can support 
```sql
select user_id from users_table where user_id < 10
union
select user_id from users_table where user_id < 10;
```

type queries. This work does not do you see a blocker on supporting this ?

File: src/backend/distributed/planner/relation_restriction_equivalence.c [hunk 4]
```diff
@@ -1052,3 +1259,93 @@ AttributeEquivalancesAreEqual(AttributeEquivalenceClass *firstAttributeEquivalan
...
+
+		if (setOperationStatement->op != SETOP_UNION)
+		{
+			return false;
+		}
+
+		return true;
+	}
+
+	return TopLevelUnionQuery(subqueryTree);
```
Comment (mtuncer) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: I did not like recursion here.  I think this is used to traverse query tree to find out the first query where there is either set operation or join. 

Can you move search logic to another walker like function and remove  recursion here ?

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 5]
```diff
@@ -2053,6 +2053,7 @@ SubquerySqlTaskList(Job *job, PlannerRestrictionContext *plannerRestrictionConte
 		subqueryTask = SubqueryTaskCreate(subquery, targetShardInterval,
 										  relationRestrictionContext, taskIdIndex);
 
+
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_physical_planner.c]: Unnecessary new line

File: src/backend/distributed/planner/relation_restriction_equivalence.c [hunk 6]
```diff
@@ -113,6 +122,187 @@ static void ListConcatUniqueAttributeClassMemberLists(AttributeEquivalenceClass
...
+		{
+			Var *targetVar = (Var *) lfirst(translatedVarCell);
+
+			childAttrNumber++;
+
+			if (targetVar->varno == relationRteIndex &&
+				targetVar->varattno == relationPartitionKey->varattno)
+			{
+				*partitionKeyIndex = childAttrNumber;
+
```
Comment (onderkalaci) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: Can't we use `targetVar ->varattno`? That seems more accurate to me.

File: src/backend/distributed/planner/relation_restriction_equivalence.c (hunk 3 above)
Comment (onderkalaci) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: > type queries. This work does not, do you see a blocker on supporting this ?

I theory, we can support that. But, I don't want to deal with !subquery code-path in this PR/release. What do you think?

File: src/backend/distributed/planner/relation_restriction_equivalence.c (hunk 6 above)
Comment (onderkalaci) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: Nope, how we use seems OK. We only need to ensure that it is actually a `Var`

File: src/backend/distributed/planner/relation_restriction_equivalence.c (hunk 3 above)
Comment (mtuncer) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: I think it won't bring any more complexity, we would just need to move down in the query funnel until there is a set operation or join.

Comment (onderkalaci) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: > I think it won't bring any more complexity, we would just need to move down in the query funnel until there is a set operation or join.

I still think that we should tackle it separately, maybe with another PR on top of this?

Comment (onderkalaci) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: Hmm, it could be that the name of the function is a bit confusing. This function is looking for a query tree such that the query may have as many upper level queries above the `UNION` query, unless none of them is a join.

Is the function comment not clear enough for that?

File: src/backend/distributed/planner/relation_restriction_equivalence.c [hunk 7]
```diff
@@ -113,6 +119,193 @@ static void ListConcatUniqueAttributeClassMemberLists(AttributeEquivalenceClass
...
+}
+
+
+/*
+ * FindTranslatedVar iterates on the appendRelList and tries to find a translated
+ * child var identified by the relation id and the relation rte index. The function
+ * returns NULL if it cannot find a translated var.
+ */
+static Var *
+FindTranslatedVar(List *appendRelList, Oid relationOid, Index relationRteIndex,
```
Comment (mtuncer) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: Is it the Var was already translated, and we are trying to find translation ? 

Function could be better named. Could you rephrase explanation.

File: src/backend/distributed/planner/relation_restriction_equivalence.c [hunk 8]
```diff
@@ -113,6 +119,193 @@ static void ListConcatUniqueAttributeClassMemberLists(AttributeEquivalenceClass
...
+
+/*
+ * SafeToPushdownTopLevelUnion returns true if all the relations are returns
+ * partition keys in the same ordinal position.
+ *
+ * Note that the function expects (and asserts) the input query to be a top
+ * level union query defined by TopLevelUnionQuery().
+ */
+bool
+SafeToPushdownTopLevelUnion(RelationRestrictionContext *restrictionContext)
```
Comment (mtuncer) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: perhaps SafeToPushdownSubqueryUnion  or SafeToPushdownUnionSubquery

File: src/backend/distributed/planner/relation_restriction_equivalence.c [hunk 9]
```diff
@@ -1052,3 +1261,93 @@ AttributeEquivalancesAreEqual(AttributeEquivalenceClass *firstAttributeEquivalan
...
+	{
+		TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+		Expr *targetExpression = targetEntry->expr;
+
+		partitionKeyTargetAttrIndex++;
+
+		if (!targetEntry->resjunk &&
+			IsPartitionColumn(targetExpression, relationPlannerParseQuery))
+		{
+			Var *targetColumn = (Var *) targetExpression;
```
Comment (mtuncer) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: Are you sure it is a Var here ?

File: src/include/distributed/multi_planner_utils.h [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+ *-------------------------------------------------------------------------
+ */
+
+#ifndef MULTI_PLANNER_UTILS_H
+#define MULTI_PLANNER_UTILS_H
+
+#include "distributed/multi_planner.h"
+
+
+extern bool SafeToPushDownSubquery(PlannerRestrictionContext *plannerRestrictionContext);
```
Comment (mtuncer) [inline src/include/distributed/multi_planner_utils.h]: This is only used inside multi_router_planner.c , declared in multi_planner_utils.h, and defined in relation_restriction_equivalance.c

It looks weird.

File: src/include/distributed/relation_restriction_equivalence.h [hunk 11]
```diff
@@ -15,8 +15,12 @@
 #include "distributed/multi_planner.h"
 
 
+extern bool SafeToPushDownSubquery(PlannerRestrictionContext *plannerRestrictionContext,
```
Comment (mtuncer) [inline src/include/distributed/relation_restriction_equivalence.h]: no need for additional declaration. it is inside  a header file now

**Comment by mtuncer** [comment]
you should check when a partition column is repeated inside target list
```sql
 select count(*) from (
        select user_id, user_id from users_table 
    union 
        select user_id, 2 from users_table) u;
```
works, but 
```sql
 select count(*) from (
        select user_id, user_id from users_table
    union 
        select 2, user_id  from users_table) u;
```
does not

**Comment by onderkalaci** [comment]
> you should check when a partition column is repeated inside target list

I'm happy to leave this as a known issue 😄  That would a very rare case, wouldn't it? Also, the solution is very simple, re-order you target lists such that partition match is at the first index?

**Comment by onderkalaci** [comment]
I'm happy to note this in the `SafeToPushdownUnionSubquery` function comment?

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 12]
```diff
@@ -2832,9 +2837,26 @@ ErrorIfContainsUnsupportedSubquery(MultiNode *logicalPlanNode,
...
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot pushdown the subquery since all leaves of "
+								   "the UNION does not include partition key at the "
+								   "same position"),
+							errdetail("Each leaf query of the UNION should return "
+									  "partition key at the same position on its "
+									  "target list.")));
+		}
+	}
+	else if (!RestrictionEquivalenceForPartitionKeys(plannerRestrictionContext))
```
Comment (mtuncer) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Does this mean we do not need to check`RestrictionEquivalenceForPartitionKeys` if the query contains top level union ?

File: src/backend/distributed/planner/multi_router_planner.c [hunk 13]
```diff
@@ -310,7 +312,7 @@ CreateInsertSelectRouterPlan(Query *originalQuery,
 		modifyTask = RouterModifyTaskForShardInterval(originalQuery, targetShardInterval,
 													  relationRestrictionContext,
 													  taskIdIndex,
-													  restrictionEquivalenceForPartitionKeys);
+													  safeToPushDownSubquery);
```
Comment (mtuncer) [inline src/backend/distributed/planner/multi_router_planner.c]: we are passing  variable `safeToPushDownSubquery ` in place of `allRelationsJoinedOnPartitionKey` perhaps we should change function api here

File: src/backend/distributed/planner/relation_restriction_equivalence.c [hunk 14]
```diff
@@ -1052,3 +1250,104 @@ AttributeEquivalancesAreEqual(AttributeEquivalenceClass *firstAttributeEquivalan
...
+		return false;
+	}
+
+	subqueryTree = rangeTableEntry->subquery;
+	setOperations = subqueryTree->setOperations;
+	if (setOperations != NULL)
+	{
+		SetOperationStmt *setOperationStatement = (SetOperationStmt *) setOperations;
+
+		if (setOperationStatement->op != SETOP_UNION)
```
Comment (mtuncer) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: may be a single line comment here stating set operation type in whole tree is checked elsewhere

File: src/test/regress/expected/multi_subquery_toplevel_union.out [hunk 15]
```diff
@@ -0,0 +1,777 @@
+--
```
Comment (mtuncer) [inline src/test/regress/expected/multi_subquery_toplevel_union.out]: since we are steering away from 'toplevel_union' we can use multi_subquery_union

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 12 above)
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: Yes, that's true, we don't need to check that.

File: src/backend/distributed/planner/multi_router_planner.c (hunk 13 above)
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_router_planner.c]: Good catch, I'll update it

File: src/backend/distributed/planner/relation_restriction_equivalence.c (hunk 14 above)
Comment (onderkalaci) [inline src/backend/distributed/planner/relation_restriction_equivalence.c]: sure, that makes sense/

File: src/test/regress/expected/multi_subquery_toplevel_union.out (hunk 15 above)
Comment (onderkalaci) [inline src/test/regress/expected/multi_subquery_toplevel_union.out]: I'll update it.
